arbnode/transaction_streamer.go

```diff
@@ -1443,24 +1457,35 @@ func (s *TransactionStreamer) submitEspressoTransactions(ctx context.Context, ig
 			log.Error("failed to get espresso submitted pos", "err", err)
 			return s.config().EspressoTxnsPollingInterval
 		}
-		bytes, _, err := arbos.ParseEspressoMsg(msg.Message)
+		txns, _, err := arbos.ParseEspressoMsg(msg.Message)
 		if err != nil {
 			log.Error("failed to parse espresso message before submitting", "err", err)
 			return s.config().EspressoTxnsPollingInterval
 		}
 
-		espressoTx := espressoTypes.Transaction{
-			Payload:   bytes[0],
-			Namespace: s.config().EspressoNamespace,
+		if s.dataSigner == nil {
+			log.Error("data signer is nil")
+			return s.config().EspressoTxnsPollingInterval
 		}
 
-		log.Info("submitting transaction to espresso using sovereign sequencer", "tx", espressoTx)
+		payloadSignature, err := s.dataSigner(crypto.Keccak256Hash(txns[0]).Bytes())
+		if err != nil {
+			log.Error("failed to sign espresso transaction", "err", err)
+			return s.config().EspressoTxnsPollingInterval
+		}
+
+		var signedPayload []byte
+
+		signedPayload = append(signedPayload, payloadSignature...)
+		signedPayload = append(signedPayload, txns[0]...)
 
 		hash, err := s.espressoClient.SubmitTransaction(ctx, espressoTypes.Transaction{
-			Payload:   bytes[0],
+			Payload:   signedPayload,
```

Comment:
I think it is a bit weird here. I think we should also sign the namespace. And maybe hotshot transaction should have a field for signature.

Reply:
Ideally we would sign the whole transaction including the namespace. However, concretely I can't think of "replaying" the tx in another namespace causing any problems, as long as we don't sign the namespace unless someone is using the same key for two namespaces. So we should leave a not about this restriction. Let's keep in mind that we will likely remove this signing feature soon anyway and not sink more time into it than necessary.

It does feel like we should have a test somewhere that does the signature recovery so that we can be somewhat sure it would work for 3rd parties that are watching the Espresso transaction stream and want to verify signatures.

Additionally I think we should add some comments to the code as to "why" we are signing the messages and that internally the signature doesn't need to verified because the batch poster receives the transactions from the sequencer. Unless that's not current, then please tell me.

cc @Sneh1999 @ImJeremyHe @zacshowa

Reply:
@sveitser @ImJeremyHe  added test and comments! Feel free to take a look when you have sometime